polynote-frontend/polynote/ui.js

```diff
@@ -918,32 +927,45 @@ export class NotebookUI {
             }
         });
 
-        socket.addMessageListener(messages.UpdateCell, (path, globalVersion, localVersion, id, edits) => {
-            if (globalVersion >= this.globalVersion) {
-                this.globalVersion = globalVersion;
-                let rebasedEdits = edits;
+        socket.addMessageListener(messages.NotebookUpdate, update => {
+            if (update.path === this.path) {
+                if (update.globalVersion >= this.globalVersion) {
+                    this.globalVersion = update.globalVersion;
 
-                if (localVersion < this.localVersion) {
-                    const editsBefore = this.editBuffer.range(localVersion, this.localVersion);
-                    rebasedEdits = messages.ContentEdit.rebaseEdits(edits, editsBefore);
-                } else if (localVersion > this.localVersion) {
-                    console.error("Local version from server is after client's. Something's wrong.")
-                }
+                    if (update.localVersion < this.localVersion) {
+                        const prevUpdates = this.editBuffer.range(update.localVersion, this.localVersion);
+                        update = messages.NotebookUpdate.rebase(update, prevUpdates);
+                    }
 
-                this.localVersion++;
 
-                const cell = this.cellUI.getCell(id);
+                    this.localVersion++;
+
+                    match(update)
+                        .when(messages.UpdateCell, (p, g, l, id, edits) => {
+                            const cell = this.cellUI.getCell(id);
+                            if (cell) {
+                                cell.applyEdits(edits);
+                                this.editBuffer.push(this.localVersion, messages);
+                            }
+                        })
+                        .when(messages.InsertCell, (p, g, l, cell, after) => {
+                            const prev = this.cellUI.getCell(after);
+                            const newCell = (prev && prev.language && prev.language !== "text")
+                                            ? new CodeCell(cell.id, cell.content, cell.language)
+                                            : new TextCell(cell.id, cell.content);
+                            this.cellUI.insertCell(newCell, after)
+                        })
+                        .when(messages.DeleteCell, (p, g, l, id) => this.cellUI.removeCell(id))
```

Comment:
if you squint real hard it almost looks like `case DeleteCell(p, g, l, id) => this.cellUI.removeCell(id)` 😛

Reply:
If you change `when` to `case` it would look even more like that!

Reply:
`case` is reserved in JS, so... no can do.

Reply:
`ccase` 😉